// @beardedframework/core/dist/services/ResponseService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class ResponseService {
    /*
     * success response
     *
     * @param { string } message
     * @param { any } data
     * @param { any } resp
     * @param { boolean } state
     *
     * @return { object }
     */
    static success(message, data, resp, state = true) {
        return resp.status(ResponseService.OK).json({
            msg: message,
            status: state,
            data: data
        });
    }
    /*
     * error response
     *
     * @param { string } message
     * @param { any } error
     * @param { any } resp
     * @param { boolean } state
     *
     * @return { object }
     */
    static error(message, error, resp, state = false) {
        return resp.status(ResponseService.INTERNAL_SERVER_ERROR).json({
            msg: message,
            status: state,
            error: error
        });
    }
    /*
     * validation needed response
     *
     * @param { string } message
     * @param { any } validation
     * @param { any } resp
     * @param { boolean } state
     *
     * @return { object }
     */
    static validationNotPassed(message, validation, resp, state = false) {
        return resp.status(ResponseService.PAYMENT_REQUIRED).json({
            msg: message,
            status: state,
            validation: validation
        });
    }
}
ResponseService.OK = 200;
ResponseService.BAD_REQUEST = 400;
ResponseService.PAYMENT_REQUIRED = 402; //used as validation error
ResponseService.FORBIDDEN = 403;
ResponseService.NOT_FOUND = 404;
ResponseService.METHOD_NOT_ALLOWED = 405;
ResponseService.INTERNAL_SERVER_ERROR = 500;
ResponseService.BAD_GATEWAY = 502;
ResponseService.SERVICE_UNAVAILABLE = 503;
exports.default = ResponseService;
